proposals/0000-splice-imports.rst

```diff
@@ -0,0 +1,1058 @@
+persistence will be disallowed by the compiler.  That is, use of a binding at a
+level other than the level at which it was defined or imported will result in a
+type error.  In particular, bindings imported using traditional ``import``
+statements may not be used inside of top-level splices, nor within quotes.
+
+For example, the following is accepted under the default ``ImplicitStagePersistence``,
+but will be rejected under ``NoImplicitStagePersistence``::
+
+   import B (foo)  -- foo :: Q Exp
+   data C = MkC
+
+   quoteC = [| MkC |]  -- Error: MkC defined at level 0 but used at level 1
+   spliceC = $( foo )  -- Error: foo imported at level 0 but used at level -1
+
+``ImplicitStagePersistence`` is enabled by default in all existing language editions.
+
+Under ``NoImplicitStagePersistence`` it is an error to use ``DeriveLift`` on a
+type unless all its definition is imported at both level 0 and level 1.
+This is discussed in more detail in the :ref:`Lift Instances <lift_instances>` section.
+
+
+
+ExplicitLevelImports
+####################
+
+The ``ExplicitLevelImports`` extension introduces two new import modifiers to
+the import syntax, ``splice`` and ``quote``, which control the level at which
+identifiers from the module are brought into scope:
+
+* A ``splice`` import of ``A`` will import all bindings of ``A`` to be used *only* at
+  level -1.
+* A ``quote`` import of ``B`` will import all bindings of ``B`` to be used
+  *only* at level 1.
+
+For example, the following is accepted under ``ExplicitLevelImports``::
+
+  import quote Foo (bar) -- bar is introduced at level 1
+  import Foo (baz) -- baz is introduced at level 0
+  import splice Foo (qux) -- qux is introduced at level -1
+
+  foo = baz [| bar |] $(qux)
+
+``ExplicitLevelImports`` implies ``NoImplicitStagePersistence``.  Thus users
+typically need only enable ``ExplicitLevelImports`` (and ``TemplateHaskell``).
+
+
+When a module uses ``TemplateHaskell`` with ``NoImplicitStagePersistence``,
+the module dependencies no longer need
+to be pessimistically compiled and loaded at compile time. Instead, the modules
+that are needed at compile-time versus runtime are determined by the explicit
+``splice`` and ``quote`` imports relative to the module being compiled.
+
+It is permitted to enable both ``ExplicitLevelImports`` and
+``ImplicitStagePersistence`` (provided the latter appears later than the former,
+so it overrides the implied ``NoImplicitStagePersistence``). This allows
+``splice`` and ``quote`` imports to be used, but ``ImplicitStagePersistence``
+still allows cross-stage persistence (and thus the compiler must still be
+pessimistically assume all modules are needed at all stages). This combination
+is supported to allow gradual migration of code bases following the change, and
+for corner cases such as code generation, where the programmer may wish to use
+the syntax of ``splice`` and ``quote`` imports without obliging the whole module
+to be level-correct.
```

Comment:
If you enable `ImplicitStagePersistence` then you get the persistence rules, so yes, then using explicit level imports will not gain you anything. It does allow you to do things like turn on `ExplicitLevelmports` globally in a project and selectively enable `ImplicitStagePersitence` on a per module basis.

Also, if you are using level imports, then the identifiers will still be brought into scope at particular levels. It is just that the persistence rules will shift these levels around in the same way. For example, if you splice import a module M then it will be made available at level -1 but then persistence might also make it available at 1, so writing `import splice M` would mean you can use things from `M` inside quotations.

The point about code generation is imagining if you are using a preprocessor or some other means of generating modules, you might want to generate imports using splice/quote even if that information is not useful in a particular project.

Reply:
> then using explicit level imports will not gain you anything. 

Could you be more explicit about what this means, in the proposal? Does that mean if *just one module* in the module graph uses ImplicitStagePersistence, then **all** modules need to be compiled at both stages? (I hope not, but that would be conservative.)

Consider again this example

```hs
module A where { a = 1 :: Int }

{-# LANGUAGE ExplicitLevelImports #-}
{-# LANGUAGE ImplicitStagePersistence #-}
module B where
import A

foo = a

bar = [| foo |]

{-# LANGUAGE ExplicitLevelImports #-}
module C where
import splice B
c :: Int
c = $(bar) 
```

Note that module B is needed both at compile-time as well as runtime, despite `bar` only being used at compile-time. That is because we end up with `c = B.foo` after splicing, so `foo` is needed at runtime.

I think what we want is as follows:

> If module `M1` uses ImplicitStagePersistence, and module `M2` imports `M1` at level n, then `M2` must also import `M1` at levels n+1,n+2,...

So if `C` imports `B` at `splice`, it must also import `B` at normal and at `quote`. If `D` imports `B` at normal, then it also needs to import `D` at `quote`, etc.

Note that this is about any modules `C` *using*/importing a module `B` using ImplicitStagePersistence. It is not about the *dependencies*/imports (such as `A`) of a module `B` using ImplicitStagePersistence.

Reply:
@sgraf812 If a module enables `ImplicitStagePersistence` then it places a requirement that the module itself and all modules in it's transitive closure are available at stages `C` and `R`. It is a very strong requirement, and a big motivation for adding the extension in the first place.

Consider this example, by inspecting the module header for C, how can you determine which stages `A` is required? Definitions from `A` are required at all levels, so you must be conservative and presume that any import can be used at any level. 

```
module A where

a = 1

{-# LANGUAGE ImplicitStagePersistence #-}
module C where

import A

-- a used at 1, requires A @ R
aq = [| a |]

-- a used at -1, requires A @ C
sq = $(lift a)
```

Reply:
Perhaps there is some confusion as `ISP` works slightly different for positive and negative persistence.

* `ISP` allows you to persist forwards identifiers from the module itself and all dependencies, therefore if `M` enables `ISP` then `M` and all its dependencies are also required at all future stages.
* `ISP` allows you to backwards persist identifiers from a modules imports, therefore if `M` enables `ISP` then
   all of `M`s dependencies are required at prior stages (but **not** `M` itself). 

This distinction is not yet made in the proposal. At the moment we say that if you enable `ISP` then you need a module for all stages which is slightly inaccurate. As there is the case that you require a module `M` at stage `R`, you enable `ISP` and this does not force you to require `M` at stage `C`.

Reply:
@mpickering the "transitive closure" part does sound a bit conservative. I would think ISP requires C and R for all *immediate* dependencies, but if those modules are non-ISP you can start to regain some invariants.

If we are too sledge-hammer with ISP, I fear we will not catch errors in non-ISP modules that happened to be used by ISP modules.

Example

```
module A where
  ..
module B where
  import quote A
  ..

module C where
  {-# LANGUAGE ImplicitStagePersistence #-}
  import B
  ..
```

We need C and R copies of `B` for `C`, but we only need `R` for `A`. This is because:

- `B` at `C` only needs `A` at `R`
- `B` at `R` only needs `A` at "super R" (level +1 which saturates to `R`

(BTW the saturation of levels to 2 phases does not seem right to me for cross compilation, but we can deal with that later.)

Reply:
Yes it does seem right that `ISP` places the requirements on immediate dependencies.